### ISSUE TITLE
Web updater stuck during patching (busy-waiting)

### ISSUE THREAD

**Issue by loafylemon** [renpy/renpy#4858]
After one of the recent updates to my project, I noticed that my server no longer effectively delivers updates for the game, despite the updates.json file being read and prepared correctly by the updater. During the update process, the program absolutely pins the CPU in terms of instructions, but the update itself does not progress. This issue occurs whether I use a remote server or a local server.

To troubleshoot the problem, I created a new debug project to eliminate any potential code-related issues and run the command `updater.update("http://127.0.0.1:3000/updates.json")`. However, the outcome remained the same. I traced the problem to the `patch` parameter, which when set to `False` allows the update to complete but is not a satisfactory solution due to the significant size of the required downloads.

To further investigate, I experimented with using different servers, such as lighttpd, python http service, and ruby. Out of these three options, only the ruby server worked as expected.

I'm curious as to why the other servers did not function correctly. I considered the possibility of missing dependencies, but after checking the documentation, I couldn't find any information regarding specific requirements that might be causing this issue.

On all failed attempts, the log file gets spammed with the following, spanning across thousands of lines.
```
downloading from http://127.0.0.1/Debug-1.1-pc.update.gz:
##################-- 93.3%
##################-- 93.3% 0.0 kBps
```

If the problem is indeed related to server configuration, it would be great to have it documented.

Pruned output: (3000+ lines)
```
running ['/home/loafy/Repositories/Debug-1.0-dists/Debug-1.0-pc/lib/py3-linux-x86_64/zsync', '-o', '/home/loafy/Repositories/Debug-1.0-dists/Debug-1.0-pc/update/pc.update.new', '-k', '/home/loafy/Repositories/Debug-1.0-dists/Debug-1.0-pc/update/pc.zsync', '-i', '/home/loafy/Repositories/Debug-1.0-dists/Debug-1.0-pc/update/pc.update', 'http://127.0.0.1/Debug-1.1-pc.zsync']
reading seed file /home/loafy/Repositories/Debug-1.0-dists/Debug-1.0-pc/update/pc.update: ***************************************************************************************
Read /home/loafy/Repositories/Debug-1.0-dists/Debug-1.0-pc/update/pc.update. Target 93.3% complete.      
downloading from http://127.0.0.1/Debug-1.1-pc.update.gz:
##################-- 93.3%
##################-- 93.3% 0.0 kBps         

downloading from http://127.0.0.1/Debug-1.1-pc.update.gz:
##################-- 93.3%
##################-- 93.3% 0.0 kBps         

downloading from http://127.0.0.1/Debug-1.1-pc.update.gz:
##################-- 93.3%
##################-- 93.3% 0.0 kBps         

downloading from http://127.0.0.1/Debug-1.1-pc.update.gz:
##################-- 93.3%
##################-- 93.3% 0.0 kBps         

downloading from http://127.0.0.1/Debug-1.1-pc.update.gz:
##################-- 93.3%
##################-- 93.3% 0.0 kBps         

downloading from http://127.0.0.1/Debug-1.1-pc.update.gz:
##################-- 93.3%
##################-- 93.3% 0.0 kBps         
```
____
**Lighttpd configuration:**

`lighttpd/1.4.63` installed via apt.

lighttpd.conf
```conf
server.document-root = "/home/loafy/Repositories/Debug-1.1-dists/"
server.errorlog = "/home/loafy/Repositories/Debug-1.1-dists/error.log"

server.port = 3000

mimetype.assign = (
  ".html" => "text/html", 
  ".txt" => "text/plain",
  ".jpg" => "image/jpeg",
  ".png" => "image/png",
  ".webp" => "image/webp",
  ".gz" => "application/gzip",
  ".zip" => "application/zip",
  ".json" => "application/json",
  ".zsync" => "application/zsync",
  ".sums" => "application/sums"
)
```

**Comment by renpytom** [comment]
Do you have reason to believe this is an issue with Ren'Py, rather than your setup? To support patching, the updater requires something very specific, which is a HTTP server that supports range queries. I'm pretty sure the Python http server doesn't, and I don't know about the other two. 

I use nginx to update Ren'Py itself.

**Comment by gstrauss** [comment]
FYI: lighttpd supports Range requests by default, unless disabled with lighttpd.conf `server.range-requests = "disable"`

Please note: Range is HTTP/1.1, and is not part of HTTP/1.0, so if you are making Range requests and using HTTP/1.0, then lighttpd will ignore Range requests for HTTP/1.0 requests and will serve the whole file, unless you set lighttpd.conf `server.feature-flags += ("http10.range" => "enable")` (which is disabled by default).

Also note: lighttpd consolidated Range requests that are close, and limits the number of ranges that can be requested in a single HTTP Range request.  The client should be able to handle if the server consolidates multiple ranges in a single HTTP Range request and returns a range greater than or equal to a (partial) set of requested ranges.  If the client sends many ranges in a single HTTP Range request, then the client should handle if only some of the ranges are fulfilled, and should issue subsequent HTTP Range requests for additional ranges.

**Comment by loafylemon** [comment]
I had no doubt that the issue was on my end. I spent a lot of time searching the internet for a solution or clue, but unfortunately, I couldn't find any. It was only after I received your message that I managed to get everything working by using nginx. I'm truly grateful that my problem is now solved, thanks to your help.

In posting my question here, I was hoping that the documentation could be updated to include specific server requirements for getting the Ren'py updater to function correctly, as this wasn't as straightforward as I was expecting it to be. I believe that even just mentioning a recommended and tested solution like nginx would be sufficient to guide and save time for others trying to set up an updater server.

If not, perhaps this thread can be helpful to others who might face a similar situation. Once again, thank you, and I apologize if this post doesn't exactly match the purpose of the issue tracker.

I'll leave it in your hands to decide whether updating the documentation is worth it or not.

**Comment by mal** [comment]
The HTTP range query requirement is already documented in the Server Requirements section of the Web Updater [docs page](https://www.renpy.org/doc/html/updater.html#server-requirements), but perhaps a list of web servers knows to support it could be useful. 🤷

**Comment by gstrauss** [comment]
I am a lighttpd developer and occasionally search through recent issues mentioning lighttpd.  Happy to help troubleshoot if you have more details like the headers sent by renpy.  lighttpd support for HTTP Range requests is advanced and RFC-compliant, so lighttpd out-of-the-box should serve any RFC-compliant HTTP Range request sent by renpy, unless lighttpd were explicitly configured to disable HTTP Range support.

It appears that renpy patching is supported only for "http:" and not for "https:".  @mal is that the case?

**Comment by mal** [comment]
> It appears that renpy patching is supported only for "http:" and not for "https:".

Correct, as I understand it that's a limitation of `zsync` which is what Ren'Py uses under-the-hood for patch updates.

**Comment by mal** [comment]
Perhaps this excerpt from the [zsync website](http://zsync.moria.org.uk/server) is relevant?
> The server hosting the full download must support HTTP Range: requests (almost all do). It should support HTTP multipart/byteranges responses, HTTP/1.1 and pipelining. Currently it must not use Transfer-Encoding: chunked (**yes, I know that makes zsync non-compliant**), and should not be generating the content dynamically.

**Comment by gstrauss** [comment]
I took a quick look at zsync.  zsync *does not* provide a fully-featured HTTP client and does not support TLS.  It also does not support HTTP/2 or HTTP/3, and does not support HTTP/1.1 Transfer-Encoding: chunked (which is a requirement for HTTP/1.1).

There are some other clever ideas in zsync, but it would have been better if zsync used a client HTTP library for HTTP, rather than writing its own -- now very dated -- implementation.

However, the likely limitation here between lighttpd and zsync is that zsync sets a limit of max 20 ranges that it might include in a single HTTP request, whereas lighttpd sets an internal limit of 10.  lighttpd sets a limit to prevent abuse or other attacks on servers by clients abusing Range.  lighttpd will sort, coalesce, and serve those 10 ranges, and the client should send a subsequent request for additional ranges.  My hunch is that zsync does not detect that the ranges were partially fulfilled, progress was made, and that zsync should make another request for additional ranges.

zsync mentions and (likely) uses HTTP/1.1 pipelining, so there is little practical difference to sending five HTTP/1.1 pipelined requests, each with 20 byte ranges, or sending ten HTTP/1.1 pipelined requests, each with 10 byte ranges.  If the byte ranges are small, they should be coalesced into a single range.  If the byte ranges are large, then the majority of time spent downloading will be sending the bytes in the ranges, and so there is little to no benefit from sending more ranges in a single HTTP request instead of sending multiple, pipelined HTTP requests.

Performance would likely be at least as good and possibly better without using HTTP/1.1 multipart/byteranges with multiple ranges, and instead using HTTP/2 streams, requesting a single range per stream.

My unverified hunch:
=> lighttpd will probably not work with the current zsync due to what lighttpd considers an excessive number of ranges.
Separately,
=> I may reach out to the zsync author to suggest libcurl or something similar.

**Comment by gstrauss** [comment]
FYI: just found this: looks like https://github.com/probonopd/zsync-curl is a fork of zsync that uses libcurl and supports HTTPS, and likely also HTTP/2 and HTTP/3, since libcurl supports those.

**Comment by mal** [comment]
Unfortunately I don't think that project or its successor(?) [zsync2](https://github.com/AppImageCommunity/zsync2) (billed as a rewrite of `zsync-curl` with the author of that project listed as a member) intend to support Windows which would be a requirement for Ren'Py. The latter does have an issue open on the topic, but decision there seems to be that while they'd consider patches, they would not support them. There are probably other considerations for Ren'Py too such as binary size, libraries and licenses involves etc, but Tom would be able to speak better to those concerns.

**Comment by gstrauss** [comment]
@mal the next version of lighttpd (lighttpd 1.4.72) updates the Range parsing algorithm to increase the number of  additional ranges allowed but only if the ranges are sorted (which zsync does), so you can update your notes to say that renpy with zsync (which sends up to 20 ranges in a single HTTP Range request)  should work with lighttpd 1.4.72 and later.

If you are looking for a more modern alternative to zsync, you might have a look at zchunk
https://github.com/zchunk/zchunk

I found an old blog post about zchunk:
https://www.jdieter.net/posts/2018/05/31/what-is-zchunk/
including its differences from zsync

**Comment by renpytom** [comment]
Thanks for the fix to lighttpd, and thanks for pointing out zchunk. I didn't know about it until now, and it looks like it's the replacement for zsync that I've been looking for.

**Comment by renpytom** [comment]
This is closed by the new updater coming in 8.2.
